proposed/2023/vs-filters.md

```diff
@@ -0,0 +1,102 @@
+# Visual Studio Filters & Sorting Options
+
+- [Jon Douglas](https://github.com/JonDouglas)
+- 2023-06-23
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+This proposal represents a common ask for more filters and sorting options in the NuGet Package Manager experience in Visual Studio. 
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+As NuGet continues to expand as a package ecosystem, there has become a dire need for being able to filter and sort packages through one's daily management in their IDE of choice. While NuGet.org supports a plethora of filters and sorts on the search experience today, we'd like to bring these similar experiences to the various tabs(browse, installed, etc) in Visual Studio to help aid various jobs-to-be-done when it comes down to package management.
+
+A recent survey in May 2023 suggests that over 31% of people do not feel like there are adequate sorting and filtering options in Visual Studio when asked. Common free-response to this question relate around the following:
+
+- "There are no filtering options"
+- "More search options would make this better"
+- "Would like to find packages by tags, authors, licenses, and more"
+- "There are no way to sort packages outside of creation date"
+- "I'm not aware of any filters but prerelease on/off"
+- "Often I'm looking for a widely-used and respected package and I have to filter and sort through random niche-market packages"
+- "I should be able to filter based on an official package source and license of a package"
+- "I manually have to use google or the console because desired results are too hard to find without filters"
+- "Filters and sorts do not match the website"
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or pseudocode to show how this proposal would look. -->
+
+When a developer is using Visual Studio to search and manage NuGet packages, they will see a new filter and sort experience. 
+
+The browse filter experience will allow the developer to filter results based on common concepts such as "frameworks", "prerelease", "owners", "download count", "license", and "created on". 
```

Comment:
Do these filters apply only to the latest package version per package ID? This is how search works today _except_ for the `prerelease` filter. Said another way, should an older package version of the same ID be returned if it matches one of these filters, instead of having the package ID entirely excluded from the result set? This would be a big change for the server side since we would need to index all package versions, not just the latest.

This is a nuanced space, so let's take a real example:
Consider the package Polly. It has the tag `Hedging` to the 8.0.0-beta.2 release (this is the latest prerelease version). This tag does not exist on 7.2.4 (this is the latest stable version).

If you search for `tag:Hedging` with prerelease included, you see Polly at the top.
![image](https://github.com/NuGet/Home/assets/94054/476de77e-1d8b-4db9-a6da-92ec55e7da52)

If you uncheck prerelease, Polly disappears entirely. That's the "added metadata" case meaning the latest version matches a filter but the non-latest does not. 
![image](https://github.com/NuGet/Home/assets/94054/832b1105-b7b8-441d-9da3-446a7489b58f)


It's a similar story for the "removed metadata" case. If you search `tag:AppFabric` + exclude prerelease, you see CacheManager.Core at the top. If you include prerelease, you no longer see CacheManager.Core at all.

All of this is because currently the search service only operates on the latest package version's metadata w.r.t to the "include prerelease" and (this is a small detail) the include SemVer 2.0.0 flag (which all recent clients set to true). So we index up to 4 package versions per package ID. 4 because there are 4 combinations of a `prerelease` flag and `include SemVer 2.0.0` flag.

For some filters, this behavior is probably fine. But what about frameworks? If the latest version of a package does not support `net46` for example, the package consumer _probably_ wants to try an older version of the package which still has that support.

I think the spec should call out whether the filtering applies only to the latest package version or to all package versions. If it's the latter then it will be a much bigger ask for the server side to implement since we'll need to index a lot more data (400K package IDs vs. 6 million package versions).

Reply:
This is my Kusto query for reference
```kql
NiPackageVersions
| where IsLatestSemVer2 or IsLatestStableSemVer2
| join kind=inner NiPackageManifests on Identity
| summarize
    LatestVersion = take_anyif(Version, IsLatestSemVer2),
    LatestStableVersion = take_anyif(Version, IsLatestStableSemVer2),
    LatestTags = take_anyif(array_sort_asc(SplitTags), IsLatestSemVer2),
    LatestSplitTags = take_anyif(array_sort_asc(SplitTags), IsLatestStableSemVer2)
    by LowerId
| where isnotempty(LatestVersion) and isnotempty(LatestStableVersion)
| where LatestVersion != LatestStableVersion
| extend Added = set_difference(LatestTags, LatestSplitTags)
| extend Removed = set_difference(LatestSplitTags, LatestTags)
| where array_length(Added) > 0 or array_length(Removed) > 0
| join kind=inner (NiPackageDownloads | distinct LowerId, TotalDownloads) on LowerId
| order by TotalDownloads desc
| project LowerId, LatestStableVersion, LatestVersion, Added, Removed, TotalDownloads
```

Reply:
It would be consistent with the current experiences as seen on screen. No experience is changing here outside the concept that you can apply additional filters and quickly sort results. So that would be latest version of a package from both the source (browse tab) and locally installed (installed/updates tab)

Frameworks would also behave this way for now. If it isn't found on stable or pre-rel channels, then the functionality just doesn't show it. Mirroring search by tfm behavior on nuget.org.

Reply:
> All of this is because **currently the search service only operates on the latest package version's metadata w.r.t to the "include prerelease" and (this is a small detail) the include SemVer 2.0.0 flag (which all recent clients set to true)**. So we index up to 4 package versions per package ID. 4 because there are 4 combinations of a `prerelease` flag and `include SemVer 2.0.0` flag.

Wow!  That makes a lot of sense as to what behavior I have seen in the past.  Unlike some, I have to maintain a lot of legacy code and often feel like Microsoft is almost deliberately tanking my experience to somehow coerce me to upgrade (as if I have not tried - it's taken 6 years to get off Framework and we're finally going to be there in a few months).

I can't tell from your comments if this is something you plan to improve, but I greatly appreciate you explaining it as a limitation.

Reply:
In the case of NuGet.org, we're not deliberately limiting the search of older version metadata to coerce users. We do it as a way to scope the amount of metadata that we index. And, perhaps most significantly, it was a decision made long ago that resulted in the UI and protocol working a specific way.

The UI and protocol all center around ID-level results, not version-level results. This does lead to smaller and perhaps more understandable results. I don't think we'd want to crowd the top results with a bunch of Newtonsoft.Json versions and I also think it's reasonable to respect the most recent wishes of the package author on what their package metadata is (i.e. the latest version metadata). Packages are immutable so there's no other way to change your metadata than to release a new version. Perhaps there's a middle ground where we only index packages when certain metadata fields change or we somehow allow users to opt into an "all version" search, but that's just not implemented today and we haven't had many complaints about it (or any in recent years that I know of). The cost would be big to change the decision at this point so we'd need a compelling reason for the user experience. Based on @JonDouglas's comment, it seems like this spec isn't the time to change course on the "indexing all vs. latest version" decision.

We have a million things on our backlog so we have to be careful about the big changes we pick up so we don't overwhelm the team. If you feel that this is something we should change, please feel free to open a proposal and gather support with upvotes (perhaps starting with a bug report on NuGet/NuGetGallery).